crates/shared/src/remaining_amounts.rs

```diff
@@ -48,19 +36,80 @@ impl Remaining {
                 order.metadata.executed_sell_amount_before_fees,
             ),
         };
+
         if order.data.partially_fillable {
-            Self::from_partially_fillable(total, executed)
+            let execution = Ratio::new_raw(
+                total
+                    .checked_sub(executed)
+                    .context("executed larger than total")?,
+                total,
+            );
+
+            let sell_amount = ratio::scalar_mul(&execution, order.data.sell_amount)
+                .context("overflow scaling sell amount for execution")?;
+            let fee_amount = ratio::scalar_mul(&execution, order.data.fee_amount)
+                .context("overflow scaling fee amount for execution")?;
+
+            let need = sell_amount
+                .checked_add(fee_amount)
+                .context("partially fillable need calculation overflow")?;
+
+            let balance = if sell_balance < need {
+                Ratio::new_raw(sell_balance, need)
+            } else {
+                ratio::one()
+            };
+
+            Ok(Self { execution, balance })
         } else {
-            Ok(Self::from_fill_or_kill(!executed.is_zero()))
+            let sell = order
+                .data
+                .sell_amount
+                .checked_add(order.data.fee_amount)
+                .context("overflow sell + fee amount")?;
+
+            let execution = if executed.is_zero() {
+                ratio::one()
+            } else {
+                ratio::zero()
+            };
+
+            let balance = if sell_balance >= sell {
+                ratio::one()
+            } else {
+                ratio::zero()
+            };
+
+            Ok(Self { execution, balance })
         }
     }
 
     /// Returns Err if the contract would error due to intermediate overflow.
     pub fn remaining(&self, total: U256) -> Result<U256> {
-        total
-            .checked_mul(self.numerator)
-            .and_then(|product| product.checked_div(self.denominator))
-            .context("overflow")
+        ratio::scalar_mul(
+            &self.balance,
+            ratio::scalar_mul(&self.execution, total)
+                .context("overflow scaling for order execution")?,
+        )
+        .context("overflow scaling for available balance")
+    }
+}
+
+mod ratio {
+    use {ethcontract::U256, num::rational::Ratio};
+
+    pub fn one() -> Ratio<U256> {
+        Ratio::new_raw(1.into(), 1.into())
+    }
+
+    pub fn zero() -> Ratio<U256> {
+        Ratio::new_raw(0.into(), 1.into())
+    }
+
+    pub fn scalar_mul(ratio: &Ratio<U256>, scalar: U256) -> Option<U256> {
+        scalar
+            .checked_mul(*ratio.numer())
+            .and_then(|product| product.checked_div(*ratio.denom()))
```

Comment:
I think this `and_then()` could probably be dropped by using `?` instead making the whole thing slightly easier to read.

Reply:
Ah right, left-over copy-pasta.